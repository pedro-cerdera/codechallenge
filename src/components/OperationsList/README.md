# OperationsList

Show all user operations in app

## Usage

```js

const operations = [
    {
      id: '1',
      icon: 'fa-plus-circle',
      title: 'Nova Transação',
      onPress: () => {},
    },
    {
      id: '2',
      icon: 'fa-list',
      title: 'Listar Transações',
      onPress: () => {},
    },
  ];
  
const Example = () => (<OperationsList operations={operations} />);
```

### Properties

| Prop                    | Description                      | Required      | Default | Format | 
| ----------------------- | ---------------------------------| ------------- | ------- | ------ |
| **`operations`**        | Array of all operations          | true.         | none    | [{id: string, icon: string, title: string, onPress: function}] |
