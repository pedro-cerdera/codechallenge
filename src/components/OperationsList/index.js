import OperationsList from './OperationsList';

export {OperationsList};
