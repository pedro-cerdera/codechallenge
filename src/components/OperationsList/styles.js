import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  container: {flexGrow: 0, width: '100%'},
  separator: {width: 10},
  itemContainer: {
    padding: 10,
    width: 100,
    height: 110,
    backgroundColor: '#FFFFFF22',
    borderRadius: 5,
    justifyContent: 'space-between',
    alignItems: 'flex-start',
  },
  itemTitle: {color: 'white'},
});
