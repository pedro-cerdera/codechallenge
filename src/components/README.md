
# Components

Here we have some components

- [OperationsList](/src/components/OperationsList/README.md)
- [Account](/src/components/Account/README.md)
