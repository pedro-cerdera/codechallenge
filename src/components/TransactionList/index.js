import TransactionList from './TransacitonList';

export {TransactionList};
