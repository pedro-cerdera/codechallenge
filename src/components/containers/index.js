export * from './SafeAreaContainer';
export * from './CardContainer';
export * from './DefaultScapingContainer';
