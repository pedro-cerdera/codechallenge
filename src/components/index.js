export * from './containers';
export * from './Account';
export * from './Icon';
export * from './OperationsList';
