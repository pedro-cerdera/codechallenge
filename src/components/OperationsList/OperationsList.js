import PropTypes from 'prop-types';
import React from 'react';
import {FlatList, TouchableOpacity, View} from 'react-native';
import {CustomText} from '../CustomText';
import {Icon} from '../Icon';
import styles from './styles';

export const OperationItem = ({icon, title, onPress}) => (
  <TouchableOpacity
    style={styles.itemContainer}
    onPress={onPress}
    testID={title}>
    <Icon name={icon} size={24} color={'white'} />
    <CustomText style={styles.itemTitle} type={'tiny'}>
      {title}
    </CustomText>
  </TouchableOpacity>
);

const OperationsList = ({operations}) => (
  <FlatList
    data={operations}
    renderItem={({item}) => <OperationItem {...item} />}
    ItemSeparatorComponent={() => <View style={styles.separator} />}
    keyExtractor={(item) => item.id}
    horizontal
    style={styles.container}
  />
);

OperationsList.propTypes = {
  operations: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      onPress: PropTypes.func.isRequired,
    }).isRequired,
  ),
};

OperationsList.defaultProps = {
  operations: [],
};

export default OperationsList;
