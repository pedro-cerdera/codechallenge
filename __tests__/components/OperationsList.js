import React from 'react';
import {FlatList} from 'react-native';
import {render, fireEvent} from 'react-native-testing-library';

import {OperationsList} from '../../src/components';
import {OperationItem} from '../../src/components/OperationsList/OperationsList';

jest.mock('../../src/components/Icon/Icon', () => 'Icon');

const mockPress = jest.fn();

const mockDataOperations = [
  {
    id: '1',
    icon: 'operation-1-icon',
    title: 'Operation-1',
    onPress: () => mockPress('operation-1'),
  },
  {
    id: '2',
    icon: 'operation-2-icon',
    title: 'Operation-2',
    onPress: () => mockPress('operation-2'),
  },
];

describe('Testing Operations', () => {
  test('Render all operations Itens correctly', () => {
    const componentTree = render(
      <OperationsList operations={mockDataOperations} />,
    );

    expect(componentTree.UNSAFE_getAllByType(FlatList).length).toBe(1);
    expect(componentTree.UNSAFE_getAllByType(OperationItem).length).toBe(
      mockDataOperations.length,
    );
  });

  test('Test click operation item', () => {
    const {getByTestId} = render(
      <OperationsList operations={mockDataOperations} />,
    );

    fireEvent.press(getByTestId('Operation-1'));
    expect(mockPress).toHaveBeenCalledWith('operation-1');
    expect(mockPress).toHaveBeenCalledTimes(1);
    expect(mockPress).not.toHaveBeenCalledWith('operation-2');
  });

  test('renders correctly', () => {
    const tree = render(
      <OperationsList operations={mockDataOperations} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
