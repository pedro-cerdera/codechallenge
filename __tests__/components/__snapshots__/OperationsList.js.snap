// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Operations renders correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "icon": "operation-1-icon",
        "id": "1",
        "onPress": [Function],
        "title": "Operation-1",
      },
      Object {
        "icon": "operation-2-icon",
        "id": "2",
        "onPress": [Function],
        "title": "Operation-2",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flexGrow": 0,
      "width": "100%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#FFFFFF22",
            "borderRadius": 5,
            "height": 110,
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": 100,
          }
        }
        testID="Operation-1"
      >
        <Icon
          color="white"
          name="operation-1-icon"
          size={24}
        />
        <Text
          style={
            Array [
              Object {
                "color": "#597183",
                "fontFamily": "Poppins-Regular",
                "fontSize": 12,
              },
              Object {
                "color": "white",
              },
            ]
          }
        >
          Operation-1
        </Text>
      </View>
      <View
        style={
          Object {
            "width": 10,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#FFFFFF22",
            "borderRadius": 5,
            "height": 110,
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": 100,
          }
        }
        testID="Operation-2"
      >
        <Icon
          color="white"
          name="operation-2-icon"
          size={24}
        />
        <Text
          style={
            Array [
              Object {
                "color": "#597183",
                "fontFamily": "Poppins-Regular",
                "fontSize": 12,
              },
              Object {
                "color": "white",
              },
            ]
          }
        >
          Operation-2
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
